src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -0,0 +1,4787 @@
+{
+	FILE          *fp;
+	const char     mnt_type_fuse[] = "fuse.daos";
+	struct mntent *fs_entry;
+	char          *pt_end;
+
+	num_dfs = 0;
+
+	fp = setmntent("/proc/self/mounts", "r");
+	if (fp == NULL) {
+		printf("Fail to open file: /proc/self/mounts\nQuit\n");
+		exit(1);
+	}
+
+	while ((fs_entry = getmntent(fp)) != NULL) {
+		if (strncmp(fs_entry->mnt_type, mnt_type_fuse, sizeof(mnt_type_fuse)) == 0) {
+			dfs_list[num_dfs].dfs_dir_hash = NULL;
+			dfs_list[num_dfs].len_fs_root  = strlen(fs_entry->mnt_dir);
+			dfs_list[num_dfs].inited       = 0;
+			dfs_list[num_dfs].pool         = NULL;
+			pt_end = stpncpy(dfs_list[num_dfs].fs_root, fs_entry->mnt_dir,
+					 DFS_MAX_PATH);
+			if ((long int)(pt_end - dfs_list[num_dfs].fs_root) >= DFS_MAX_PATH) {
+				printf("dfs_list[num_dfs].fs_root[] is not large enough.\nQuit\n");
+				exit(1);
+			}
+			num_dfs++;
+		}
+	}
+
+	endmntent(fp);
+}
+
+static int
+retrieve_handles_from_fuse(int idx)
+{
+	struct dfuse_hs_reply hs_reply;
+	int                   fd, cmd, rc, errno_saved;
+	d_iov_t               iov = {};
+	char                  *buff = NULL;
+	size_t                buff_size;
+
+	fd = libc_open(dfs_list[idx].fs_root, O_RDONLY | O_DIRECTORY | O_NOFOLLOW);
+	if (fd < 0) {
+		if (bLog)
+			printf("Error: failed to open dir: %s\n", dfs_list[idx].fs_root);
+		return (-1);
+	}
+
+	cmd = _IOR(DFUSE_IOCTL_TYPE, DFUSE_IOCTL_REPLY_SIZE, struct dfuse_hs_reply);
+	rc  = ioctl(fd, cmd, &hs_reply);
+	if (rc != 0) {
+		if (errno == EPERM)
+			errno = EACCES;
+		if (bLog)
+			printf("Error in querying size info from dfuse with ioctl().\n");
+		goto err;
+	}
+
+	/* To determine the size of buffer we need to accommodate the data from fuse */
+	buff_size = max(hs_reply.fsr_pool_size, hs_reply.fsr_cont_size);
+	buff_size = max(buff_size, hs_reply.fsr_dfs_size);
+	D_ALLOC(buff, buff_size);
+	if (buff == NULL)
+		goto err;
+
+	iov.iov_buf = buff;
+	cmd = _IOC(_IOC_READ, DFUSE_IOCTL_TYPE, DFUSE_IOCTL_REPLY_POH, hs_reply.fsr_pool_size);
+	rc  = ioctl(fd, cmd, iov.iov_buf);
+	if (rc != 0) {
+		if (bLog)
+			printf("Fail to query pool handle from dfuse with ioctl().\n");
+		goto err;
+	}
+	iov.iov_buf_len = hs_reply.fsr_pool_size;
+	iov.iov_len     = iov.iov_buf_len;
+	rc              = daos_pool_global2local(iov, &dfs_list[idx].poh);
+	if (rc != 0) {
+		if (bLog)
+			printf("Fail to create pool handle in daos_pool_global2local().\n");
+		errno = rc;
+		goto err;
+	}
+
+	cmd = _IOC(_IOC_READ, DFUSE_IOCTL_TYPE, DFUSE_IOCTL_REPLY_COH, hs_reply.fsr_cont_size);
+	rc  = ioctl(fd, cmd, iov.iov_buf);
+	if (rc != 0) {
+		if (bLog)
+			printf("Fail to query container handle from dfuse with ioctl().\n");
+		goto err;
+	}
+	iov.iov_buf_len = hs_reply.fsr_cont_size;
+	iov.iov_len     = iov.iov_buf_len;
+	rc              = daos_cont_global2local(dfs_list[idx].poh, iov, &dfs_list[idx].coh);
+	if (rc != 0) {
+		if (bLog)
+			printf("Fail to create container handle in daos_pool_global2local().\n");
+		errno = rc;
```

Comment:
These functions return daos error codes, you need to translate to system error codes before returning.

Reply:
Thank you! daos_der2errno() is called to translate daso error to errno.